csharp/src/Drivers/BigQuery/BigQueryConnection.cs

```diff
@@ -0,0 +1,954 @@
+            string dataType = index == -1 ? type : type.Substring(0, index);
+
+            return GetFieldBuilder(name, type, dataType, index);
+        }
+
+        private Field.Builder GetFieldBuilder(string name, string type, string dataType, int index)
+        {
+            Field.Builder fieldBuilder = new Field.Builder();
+            fieldBuilder.Name(name);
+
+            switch (dataType)
+            {
+                case "INTEGER" or "INT64":
+                    return fieldBuilder.DataType(Int64Type.Default);
+                case "FLOAT" or "FLOAT64":
+                    return fieldBuilder.DataType(DoubleType.Default);
+                case "BOOL" or "BOOLEAN":
+                    return fieldBuilder.DataType(BooleanType.Default);
+                case "STRING" or "GEOGRAPHY":
+                    return fieldBuilder.DataType(StringType.Default);
+                case "BYTES":
+                    return fieldBuilder.DataType(BinaryType.Default);
+                case "DATETIME":
+                    return fieldBuilder.DataType(TimestampType.Default);
+                case "TIMESTAMP":
+                    return fieldBuilder.DataType(TimestampType.Default);
+                case "TIME":
+                    return fieldBuilder.DataType(Time64Type.Default);
+                case "DATE":
+                    return fieldBuilder.DataType(Date64Type.Default);
+                case "RECORD" or "STRUCT":
+                    string fieldRecords = type.Substring(index + 1);
+                    fieldRecords = fieldRecords.Remove(fieldRecords.Length - 1);
+                    List<Field> nestedFields = new List<Field>();
+
+                    foreach (string record in fieldRecords.Split(','))
+                    {
+                        string fieldRecord = record.Trim();
+                        string fieldName = fieldRecord.Split(' ')[0];
+                        string fieldType = fieldRecord.Split(' ')[1];
+                        nestedFields.Add(SchemaFieldGenerator(fieldName, fieldType).Build());
+                    }
+
+                    return fieldBuilder.DataType(new StructType(nestedFields));
+                case "NUMERIC" or "DECIMAL":
+                    ParsedDecimalValues values128 = ParsePrecisionAndScale(type);
+                    return fieldBuilder.DataType(new Decimal128Type(values128.Precision, values128.Scale));
+                case "BIGNUMERIC" or "BIGDECIMAL":
+                    ParsedDecimalValues values256 = ParsePrecisionAndScale(type);
+                    return fieldBuilder.DataType(new Decimal256Type(values256.Precision, values256.Scale));
+                case "ARRAY":
+                    string arrayType = type.Substring(dataType.Length).Replace("<", "").Replace(">", "");
+                    return GetFieldBuilder(name, type, arrayType, index);
+
+                default: throw new InvalidOperationException();
+            }
+        }
+
+        private class ParsedDecimalValues
```

Comment:
struct?

Reply:
On the other hand, looking at the decimal typename parsing code maybe it's too early to push on microoptimization...